File: practice-projects/gin-ddd/internal/infrastructure/persistence/order/order_converter.go
package order

import (
	orderModel "gin-ddd/internal/domain/model/order"
)

// toDO 将订单领域模型转换为数据对象,所有字段映射在编译期可见。
func toDO(o *orderModel.Order) *OrderDO {
	if o == nil {
		return nil
	}
	return &OrderDO{
		ID:          o.OrderID,
		OrderNo:     o.OrderNo,
		UserID:      o.UserID,
		TotalAmount: o.TotalAmount,
		Status:      string(o.Status),
		CreatedAt:   o.CreatedAt,
		UpdatedAt:   o.UpdatedAt,
	}
}

// toModel 将数据对象还原为领域模型,经聚合根 Restore 工厂构建,不触发领域事件。
func toModel(do *OrderDO) *orderModel.Order {
	if do == nil {
		return nil
	}
	return orderModel.Restore(
		do.ID,
		do.OrderNo,
		do.UserID,
		do.TotalAmount,
		orderModel.OrderStatus(do.Status),
		do.CreatedAt,
		do.UpdatedAt,
	)
}

// toModels 批量转换。
func toModels(dos []*OrderDO) []*orderModel.Order {
	if dos == nil {
		return nil
	}
	out := make([]*orderModel.Order, 0, len(dos))
	for _, do := range dos {
		out = append(out, toModel(do))
	}
	return out
}

File: practice-projects/springboot4ddd/src/main/java/com/github/microwind/springboot4ddd/infrastructure/repository/order/OrderConverter.java
package com.github.microwind.springboot4ddd.infrastructure.repository.order;

import com.github.microwind.springboot4ddd.domain.model.order.Order;

import java.util.List;
import java.util.stream.Collectors;

/**
 * 订单领域模型 ↔ 数据对象 显式转换器
 *
 * <p>放在 infrastructure 层，避免领域模型反向依赖持久化结构。
 * 不使用 {@code BeanUtils.copyProperties}，所有字段映射在编译期可见，
 * 字段调整时编译器会立刻报错。
 *
 * @author jarry
 * @since 1.0.0
 */
public final class OrderConverter {

    private OrderConverter() {
    }

    public static OrderDO toDO(Order order) {
        if (order == null) {
            return null;
        }
        return OrderDO.builder()
                .id(order.getId())
                .orderNo(order.getOrderNo())
                .userId(order.getUserId())
                .totalAmount(order.getTotalAmount())
                .status(order.getStatus().name())
                .createdAt(order.getCreatedAt())
                .updatedAt(order.getUpdatedAt())
                .build();
    }

    public static Order toModel(OrderDO orderDO) {
        if (orderDO == null) {
            return null;
        }
        return Order.restore(
                orderDO.getId(),
                orderDO.getOrderNo(),
                orderDO.getUserId(),
                orderDO.getTotalAmount(),
                Order.OrderStatus.valueOf(orderDO.getStatus()),
                orderDO.getCreatedAt(),
                orderDO.getUpdatedAt()
        );
    }

    public static List<Order> toModelList(List<OrderDO> dos) {
        if (dos == null) {
            return null;
        }
        return dos.stream()
                .map(OrderConverter::toModel)
                .collect(Collectors.toList());
    }
}
